org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCaffeineCache.java

```diff
@@ -0,0 +1,165 @@
+                .recordStats()
+                .build();
+
+        dfsBlockCache = Caffeine.newBuilder()
+                .maximumSize(cacheEntrySize)
+                .expireAfterAccess(cacheConfig.getPackFileExpireSeconds(), TimeUnit.SECONDS)
+                .recordStats()
+                .build();
+    }
+
+    boolean shouldCopyThroughCache(long length) {
+        return length <= maxStreamThroughCache;
+    }
+
+    DfsPackFile getOrCreate(DfsPackDescription description, DfsPackKey key) {
+        DfsPackFile packFile = packFileCache.getIfPresent(description);
+        if (packFile != null && !packFile.invalid()) {
+            return packFile;
+        }
+
+        DfsPackFile newPackFile = new DfsPackFile(this, description, key != null ? key : new DfsPackKey());
+        packFileCache.put(description, newPackFile);
+        return newPackFile;
+    }
+
+    int getBlockSize() {
+        return blockSize;
+    }
+
+    DfsBlock getOrLoad(DfsPackFile pack, long position, DfsReader dfsReader) throws IOException {
+        final long requestedPosition = position;
+        position = pack.alignToBlock(position);
+        DfsPackKey key = pack.key;
+
+        Ref<DfsBlock> loadedBlockRef = dfsBlockCache.get(new DfsPackKeyWithPosition(key, position), keyWithPosition -> {
+            try {
+                DfsBlock loadedBlock = pack.readOneBlock(keyWithPosition.getPosition(), dfsReader);
+                return new Ref(keyWithPosition.getDfsPackKey(), keyWithPosition.getPosition(), blockSize, loadedBlock);
+            } catch (IOException e) {
+                return null;
+            }
+        });
+
+        if (loadedBlockRef != null) {
+            DfsBlock loadedBlock = loadedBlockRef.get();
+            if (loadedBlock != null && loadedBlock.contains(key, position)) {
+                return loadedBlock;
+            }
+        }
+
+        return getOrLoad(pack, requestedPosition, dfsReader);
+    }
+
+    void put(DfsBlock block) {
+        put(block.pack, block.start, block.size(), block);
+    }
+
+    public <T> Ref<T> put(DfsPackKey key, long position, int size, T value) {
+        return dfsBlockCache.get(new DfsPackKeyWithPosition(key, position),
+                keyWithPosition -> new Ref(key, position, size, value));
+    }
+
+    boolean contains(DfsPackKey key, long position) {
+        return get(key, position) != null;
+    }
+
+    @SuppressWarnings("unchecked")
+    <T> T get(DfsPackKey key, long position) {
+        Ref<T> blockCache = dfsBlockCache.getIfPresent(new DfsPackKeyWithPosition(key, position));
+        return blockCache != null ? blockCache.get() : null;
+    }
+
+    void remove(DfsPackFile pack) {
+        packFileCache.invalidate(pack.getPackDescription());
+    }
+
+    void cleanUp() {
+        packFileCache.invalidateAll();
+        dfsBlockCache.invalidateAll();
+    }
+
+    private static final class DfsPackKeyWithPosition {
```

Comment:
A hash map uses the key's equals and hashCode to locate and store an entry. If two equivalent but not equal keys are used, the map will consider them pointing to distinct entries. A cache is built on a hash map.

```
DfsPackKeyWithPosition key1 = new DfsPackKeyWithPosition(packKey, 100);
DfsPackKeyWithPosition key2 = new DfsPackKeyWithPosition(packKey, 100);
assert key1.equals(key)
assert key.hashCode() == key2.hashCode()
```